cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java

```diff
@@ -455,7 +455,11 @@ private void initializeControllers(long startMillis) {
             }
             completed.set(false);
             try {
-              if (createControllerIfNeeded(runRecordDetail)) {
+              // Controller only needs to be created for program runs in RUNNING or SUSPENDED state.
+              // Program runs in PENDING and STARTING state will eventually start and controller will be created later.
+              if (runRecordDetail.getStatus() != ProgramRunStatus.PENDING
```

Comment:
shall we put those additional condition into createControllerIfNeeded , as they are part of the criteria whether the controller needs to be created.

Reply:
Great suggestion. Done!